packages/@aws-cdk/aws-ec2/lib/launch-template.ts

```diff
@@ -0,0 +1,688 @@
+   */
+  readonly interruptionBehavior?: SpotInstanceInterruption;
+
+  /**
+   * Maximum hourly price you're willing to pay for each Spot instance. The value is given
+   * in dollars. ex: 0.01 for 1 cent per hour, or 0.001 for one-tenth of a cent per hour.
+   *
+   * @default Maximum hourly price will default to the on-demand price for the instance type.
+   */
+  readonly maxPrice?: number;
+
+  /**
+   * The Spot Instance request type.
+   *
+   * If you are using Spot Instances with an Auto Scaling group, use one-time requests, as the
+   * Amazon EC2 Auto Scaling service handles requesting new Spot Instances whenever the group is
+   * below its desired capacity.
+   *
+   * @default One-time spot request.
+   */
+  readonly requestType?: SpotRequestType;
+
+  /**
+   * The end date of the request. For a one-time request, the request remains active until all instances
+   * launch, the request is canceled, or this date is reached. If the request is persistent, it remains
+   * active until it is canceled or this date and time is reached.
+   *
+   * @default The default end date is 7 days from the current date.
+   */
+  readonly validUntil?: Expiration;
+};
+
+/**
+ * The types of resource tags that can be set on a launch template.
+ */
+export enum LaunchTemplateTagResourceType {
+  // dev-note: Full list at https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-ec2-launchtemplate-tagspecification.html#cfn-ec2-launchtemplate-tagspecification-resourcetype
+  // These are the only ones from the list that presently support tag on create.
+
+  /**
+   * EC2 Instances.
+   */
+  INSTANCE = 'instance',
+
+  /**
+   * EBS Volumes.
+   */
+  VOLUME = 'volume',
+}
+
+/**
+ * Interface for defining the tag specification of a LaunchTemplate.
+ */
+export interface LaunchTemplateTagSpecification {
+  /**
+   * The types of resources that are tagged with the given tags. Note that there are
+   * restrictions regarding which resources can be tagged when created. See the following
+   * for additional information.
+   *
+   * @see https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-ec2-launchtemplate-tagspecification.html#cfn-ec2-launchtemplate-tagspecification-resourcetype
+   */
+  readonly resourceTypes: LaunchTemplateTagResourceType[];
+
+  /**
+   * The tags to apply to the resources.
+   */
+  readonly tags: CfnTag[];
+}
+
+/**
+ * Properties of a LaunchTemplate.
+ */
+export interface LaunchTemplateProps {
+  /**
+   * Name for this launch template.
+   *
+   * @default Automatically generated name
+   */
+  readonly launchTemplateName?: string;
+
+  /**
+   * Type of instance to launch.
+   *
+   * @default - This Launch Template does not specify a default Instance Type.
+   */
+  readonly instanceType?: InstanceType;
+
+  /**
+   * The AMI that will be used by instances.
+   *
+   * @default - This Launch Template does not specify a default AMI.
+   */
+  readonly machineImage?: IMachineImage;
```

Comment:
A reasonable argument could be probably made for making this a required property; it would allow us to always provide a default userData and mimic the `osType` + `userData` interface that is available on the Instance and AutoScalingGroup L2s. 

Of the four uses of launch templates that are identified above, the Instance and AutoScalingGroup could both potentially provide a value for a machineImage outside of a particular launch template, but the later needs to provide the override using a launch template.

I'm not familiar enough with launch template use-cases to make a strong argument one way or the other; not sure if there are any uses that get broken if we make this required.

Reply:
Let's do it like this and see where it breaks.

I'll 100% give you that the moment we make this required, someone will come by and say "but I use Launch Templates to give all my instances the same networking configuration but then configure the AMI on the `Instance` class".

Reply:
True enough. Let's leave it as optional.